ruma-common/src/push.rs

```diff
@@ -80,6 +85,97 @@ impl Ruleset {
             AnyPushRule::Sender(r) => self.sender.insert(r),
         }
     }
+
+    /// Get the push actions that apply to this event.
+    ///
+    /// Returns `None` if no push action applies.
+    ///
+    /// # Arguments
+    ///
+    /// * `event` - The raw JSON of a room message event.
+    /// * `context` - The context of the message and room at the time of the event.
+    pub fn get_actions<'a, T>(
+        &'a self,
+        event: &Raw<T>,
+        context: &PushConditionRoomCtx,
+    ) -> Option<impl Iterator<Item = &'a Action>> {
+        let event_map = &flatten_json(event);
+
+        for rule in self.override_.iter().filter(|r| r.enabled) {
+            if rule.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.content.iter().filter(|r| r.enabled) {
+            let condition = PushCondition::EventMatch {
+                key: "content.body".to_string(),
+                pattern: rule.pattern.clone(),
+            };
+
+            if condition.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.room.iter().filter(|r| r.enabled) {
+            let condition = PushCondition::EventMatch {
+                key: "room_id".to_string(),
+                pattern: rule.rule_id.clone(),
+            };
+
+            if condition.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.sender.iter().filter(|r| r.enabled) {
+            let condition = PushCondition::EventMatch {
+                key: "sender".to_string(),
+                pattern: rule.rule_id.clone(),
+            };
+
+            if condition.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.underride.iter().filter(|r| r.enabled) {
+            if rule.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+
+        None
+    }
+}
+
+fn flatten_json<T>(json: T) -> BTreeMap<String, String>
+where
+    T: Serialize,
+{
+    let flat = &mut BTreeMap::new();
+
+    flatten_value(to_json_value(json).expect("Could not parse JSON"), flat, "".to_string());
+
+    flat.to_owned()
+}
+
+fn flatten_value(value: JsonValue, map: &mut BTreeMap<String, String>, path: String) {
+    match value {
+        JsonValue::Object(fields) => {
+            for (key, value) in fields {
+                let path = if path.is_empty() { key } else { format!("{}.{}", path, key) };
+                flatten_value(value, map, path);
+            }
+        }
+        JsonValue::String(s) => {
+            if map.insert(path.clone(), s).is_some() {
+                warn!("Duplicate key in flattened JSON: {}", path);
+            };
+        }
+        _ => {
+            if map.insert(path.clone(), value.to_string()).is_some() {
```

Comment:
I'm not sure whether stringifying arrays and numbers is the right thing to do. I think we should probably skip them entirely (i.e. do nothing in the `_` match arm).

Reply:
I agree about arrays, because I don't see how they could be matched with the current keys, but I thought maybe a number or boolean could be matched.

Reply:
Hm, sure. In that case please make that more explicit with `value @ JsonValue::Number(_) | value @ JsonValue::Boolean(_)` and `JsonValue::Array(_)` branches.